File: lib/JFFNetwork/Errors/JStreamError.m
#import "JStreamError.h"

@implementation JStreamError
{
@private
    CFStreamError _rawError;
    id<NSCopying> _context;
}

- (instancetype)init
{
    [self doesNotRecognizeSelector:_cmd];
    return nil;
}

- (instancetype)initWithDescription:(NSString *)description
                             domain:(NSString *)domain
                               code:(NSInteger)code
{
    [self doesNotRecognizeSelector:_cmd];
    return nil;
}

- (instancetype)initWithStreamError:(CFStreamError)rawError
                            context:(id<NSCopying>)context
{
    NSString *domain_ = [[NSString alloc] initWithFormat: @"com.just_for_fun.library.network.CFError(%ld)", rawError.domain];
    
    self = [super initWithDescription:NSLocalizedString(@"JNETWORK_CF_STREAM_ERROR", nil)
                               domain:domain_
                                 code:rawError.error];
    
    if (nil == self) {
        
        return nil;
    }
    
    _rawError = rawError;
    _context  = context;
    
    return self;
}

- (instancetype)copyWithZone:(NSZone *)zone
{
    JStreamError *copy = [super copyWithZone:zone];
    
    if (copy) {
        
        copy->_rawError = _rawError;
        copy->_context  = [_context copyWithZone:zone];
    }
    
    return copy;
}

- (NSString *)errorLogDescription
{
    return [[NSString alloc] initWithFormat:@"%@ nativeError domain:%ld error_code:%d context:%@",
            [self localizedDescription],
            _rawError.domain,
            (int)_rawError.error,
            _context];
}

@end

File: iAsync.Swift/lib/JNetwork/JNetwork/NSError/JStreamError.swift
//
//  JStreamError.swift
//  Wishdates
//
//  Created by Vladimir Gorbenko on 18.08.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

import JUtils

public class JStreamError : JNetworkError {
    
    let streamError: CFStreamError
    private let context: NSCopying
    
    required public init(streamError: CFStreamError, context: NSCopying) {
        
        self.streamError = streamError
        self.context     = context
        
        let domain = "com.just_for_fun.library.network.CFError(\(streamError.domain))"
        let description = "JNETWORK_CF_STREAM_ERROR"
        
        super.init(description: description, domain: domain, code: Int(streamError.error))//TODO streamError.error)
    }
    
    required public init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public override func copyWithZone(zone: NSZone) -> AnyObject {
        
        return self.dynamicType(streamError: streamError, context: context)
    }
    
    public override var errorLogDescription: String {
        
        return "\(self.dynamicType) : \(localizedDescription) nativeError domain:\(streamError.domain) error_code:\(streamError.error) context:\(context)"
    }
}
